pkg/avfoundation/avfoundation_darwin.go

```diff
@@ -118,8 +118,8 @@ func (rc *ReadCloser) dataCb(data []byte) {
 		return
 	}
 	select {
+	// Use the Done channel to avoid waiting for new data from closed camera
 	case <-rc.cancelCtx.Done():
-		close(rc.dataChan)
 	case rc.dataChan <- data:
```

Comment:
Won't it cause `send to closed channel` when `Close()` while running?

Reply:
You are right, if the context is marked as done, the result of this select is random, can be both ok or panic.
I probably just had lot of luck when tested it ^^

I will search a better way to handle this.

Reply:
Ok, after some tests, it's not luck.

We cancel the context before closing the channel, so the `case <-rc.cancelCtx.Done():` branch will always be triggered before the write on closed channel.

I can unregister the callback before closing the channel if you want but it shouldn't change any thing since we also check for the context not being canceled at the start of the callback.

Reply:
I wrote a simplified test and it seems panic at low probability.
```go
package main

import (
  "context"
  "sync"
  "testing"
)

func TestMain(t *testing.T) {
  const n = 1000000
  var nPanic int
  for i := 0; i < n; i++ {
    func() {
      defer func() {
        if r := recover(); r != nil {
          t.Errorf("Panic: %v", r)
          nPanic++
        } 
      }()
      ch := make(chan struct{})
      ctx, cancel := context.WithCancel(context.Background())

      var wg sync.WaitGroup

      wg.Add(1)
      go func() {
        cancel()
        close(ch)
        wg.Done()
      }()

      select {
      case ch <- struct{}{}:
      case <-ctx.Done():
      }
      wg.Wait()
    }()
  }
  if nPanic > 0 {
    t.Fatalf("panic %d/%d", nPanic, n)
  } 
}
```
Around 20/1000000 on my environment

Reply:
I think this logic is used all over the pion code.
It would be ok to merge this for now to avoid larger problem, but better to have safe logic.

Reply:
Ok, will try to avoid this.

My idea is to use a WaitingGroup to be sure every callbacks have returned and that I can safely close the channel.
What do you think ?

Reply:
Sounds good to me!